### ISSUE TITLE
wolfssl and mbedtls etc support?

### ISSUE THREAD

**Issue by laoshaw** [mortzdk/Websocket#7]
Can it be ssl-neutral so it supports other ssl libraries in addition to openssl? e.g. wolfssl, mbedtls, boringssl,libressl

**Comment by mortzdk** [comment]
It would require some rewriting/refactoring of the code. It would actually be nice to have all SSL specific logic placed under the same file. Then it would also be a lot easier to use other libraries.

I have not really looked at alternatives for the SSL part yet. Currently I do not know if there is major API differences between the different implementation, but if there is not, then a bit of refactoring should allow usage of more of them. I do however not see how this can be done fully neutral, as I guess the API's are somewhat different.

**Comment by mortzdk** [comment]
@laoshaw, see https://github.com/mortzdk/Websocket/tree/feature/ssl-neutral.

This is my current refactoring of the SSL part of the server.

It introduces support for wolfssl, boringssl and libressl.

Now to compile with another SSL library than openssl:

**wolfSSL**
`apt-get install libwolfssl-dev`
`make SSL=WOLFSSL`

**LibreSSL**
Compile from source: https://github.com/libressl-portable/portable
`make SSL=LIBRESSL SSL_LIBRARY_PATH=/path/to/libressl`

**BoringSSL**
Compile from source: https://github.com/google/boringssl
`make SSL=BORINGSSL SSL_LIBRARY_PATH=/path/to/boringssl`

**Comment by laoshaw** [comment]
I failed to build `make SSL=WOFLSSL` on ubuntu 20.04:

```
gcc -O3 -funroll-loops -DNDEBUG -fno-exceptions -fPIC -fstack-protector -fvisibility=hidden -march=native -MMD -pedantic -pedantic-errors -pipe -W -Wall -Werror -Wno-unused-parameter -Wno-unused-function -Wno-strict-aliasing -Wno-overlength-strings -Wformat -Wformat-security -Wformat-nonliteral -Winit-self -Winline -Wl,-z,relro -Wl,-z,now -Wmultichar -Wpointer-arith -Wreturn-type -Wsign-compare -Wuninitialized -D_DEFAULT_SOURCE -DWSS_SERVER_VERSION=\"v2.0.7\" -DUSE_RPMALLOC   -DUSE_WOLFSSL  -std=c11 -I/home/shawn/Websocket/include -I/home/shawn/Websocket/src -I/home/shawn/Websocket/extensions -I/home/shawn/Websocket/subprotocols -c /home/shawn/Websocket/src/ssl.c -o /home/shawn/Websocket/build/ssl.o
In file included from /usr/include/wolfssl/ssl.h:33,
                 from /home/shawn/Websocket/src/ssl.c:22:
/usr/include/wolfssl/wolfcrypt/settings.h:1995:14: error: #warning "For timing resistance / side-channel attack prevention consider using harden options" [-Werror=cpp]
 1995 |             #warning "For timing resistance / side-channel attack prevention consider using harden options"
      |              ^~~~~~~
/home/shawn/Websocket/src/ssl.c: In function ‘WSS_http_ssl’:
/home/shawn/Websocket/src/ssl.c:210:14: error: implicit declaration of function ‘TLS_method’ [-Wimplicit-function-declaration]
  210 |     method = TLS_method();
      |              ^~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:210:12: error: assignment to ‘WOLFSSL_METHOD *’ {aka ‘struct WOLFSSL_METHOD *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  210 |     method = TLS_method();
      |            ^
/home/shawn/Websocket/src/ssl.c:247:46: error: ‘SSL_OP_SINGLE_ECDH_USE’ undeclared (first use in this function)
  247 |     wolfSSL_CTX_set_options(server->ssl_ctx, SSL_OP_SINGLE_ECDH_USE);
      |                                              ^~~~~~~~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:247:46: note: each undeclared identifier is reported only once for each function it appears in
/home/shawn/Websocket/src/ssl.c:256:43: error: ‘SSL_MODE_ENABLE_PARTIAL_WRITE’ undeclared (first use in this function)
  256 |     wolfSSL_CTX_set_mode(server->ssl_ctx, SSL_MODE_ENABLE_PARTIAL_WRITE);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:262:43: error: ‘SSL_MODE_RELEASE_BUFFERS’ undeclared (first use in this function)
  262 |     wolfSSL_CTX_set_mode(server->ssl_ctx, SSL_MODE_RELEASE_BUFFERS);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:266:47: error: ‘SSL_OP_NO_COMPRESSION’ undeclared (first use in this function)
  266 |         wolfSSL_CTX_set_mode(server->ssl_ctx, SSL_OP_NO_COMPRESSION);
      |                                               ^~~~~~~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:270:43: error: ‘SSL_OP_CIPHER_SERVER_PREFERENCE’ undeclared (first use in this function)
  270 |     wolfSSL_CTX_set_mode(server->ssl_ctx, SSL_OP_CIPHER_SERVER_PREFERENCE);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:274:46: error: ‘SSL_OP_NO_TICKET’ undeclared (first use in this function)
  274 |     wolfSSL_CTX_set_options(server->ssl_ctx, SSL_OP_NO_TICKET);
      |                                              ^~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:275:46: error: ‘SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION’ undeclared (first use in this function)
  275 |     wolfSSL_CTX_set_options(server->ssl_ctx, SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION);
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c:289:13: error: implicit declaration of function ‘SSL_CTX_set_cipher_list’; did you mean ‘wolfSSL_CTX_set_cipher_list’? [-Wimplicit-function-declaration]
  289 |             SSL_CTX_set_cipher_list(server->ssl_ctx, list);
      |             ^~~~~~~~~~~~~~~~~~~~~~~
      |             wolfSSL_CTX_set_cipher_list
/home/shawn/Websocket/src/ssl.c:295:62: error: ‘cipher_list’ undeclared (first use in this function)
  295 |         wolfSSL_CTX_set_cipher_list(server->ssl_ctx, (char *)cipher_list);
      |                                                              ^~~~~~~~~~~
/home/shawn/Websocket/src/ssl.c: In function ‘WSS_ssl_read’:
/home/shawn/Websocket/src/ssl.c:352:11: error: implicit declaration of function ‘SSL_get_error’; did you mean ‘wolfSSL_get_error’? [-Wimplicit-function-declaration]
  352 |     err = SSL_get_error(session->ssl, n);
      |           ^~~~~~~~~~~~~
      |           wolfSSL_get_error
/home/shawn/Websocket/src/ssl.c: In function ‘WSS_ssl_handshake’:
/home/shawn/Websocket/src/ssl.c:520:11: error: implicit declaration of function ‘wolfSSL_SSL_do_handshake’ [-Wimplicit-function-declaration]
  520 |     ret = wolfSSL_SSL_do_handshake(session->ssl);
      |           ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make: *** [Makefile:201: /home/shawn/Websocket/build/ssl.o] Error 1

```

**Comment by mortzdk** [comment]
@laoshaw thanks for the replay.

I deleted an include by mistake in some refactoring....

Pull the branch and try again - now it should be working :)

**Comment by laoshaw** [comment]
Builds fine and runs well. Thanks for the quick fix.

**Comment by mortzdk** [comment]
I'll do a new release with this feature one of the following days.

You welcome.

**Comment by mortzdk** [comment]
This is now a part of release 2.1.0
